File: word-ladder/word-ladder.java
class Solution {
    private boolean reachable(String s1, String s2) {
        if (s1.length() != s2.length()) {
            return false;
        }
        boolean hasDiff = false;
        for (int i = 0; i < s1.length(); i++) {
            if (s1.charAt(i) != s2.charAt(i)) {
                if (hasDiff) {
                    return false;
                } else {
                    hasDiff = true;
                }
            }
        }
        return true;
    }
    
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        List<List<Integer>> adjList = new ArrayList<>();
        int beginIndex = wordList.indexOf(beginWord);
        int endIndex = -1;
        
        if (beginIndex == -1) {
            beginIndex = wordList.size();
            wordList.add(beginWord);
        }
        
        for (int i = 0; i < wordList.size(); i++) {
            String word = wordList.get(i);
            if (word.equals(endWord)) {
                endIndex = i;
            }
            List<Integer> neighbours = new ArrayList<>();
            for (int j = 0; j < wordList.size(); j++) {
                if (reachable(word, wordList.get(j))) {
                    neighbours.add(j);
                }
            }
            adjList.add(neighbours);
        }
        
        if (endIndex == -1) {
            return 0;
        }
        
        Queue<Integer> queue = new ArrayDeque<>();
        Set<Integer> visited = new HashSet<>();
        queue.offer(beginIndex);
        visited.add(beginIndex);
        
        int count = 1;
        while (!queue.isEmpty()) {
            int size = queue.size();
            count++;
            for (int i = 0; i < size; i++) {
                int curr = queue.poll();
                for (int neighborIndex : adjList.get(curr)) {
                    if (neighborIndex == endIndex) {
                        return count;
                    }
                    if (visited.contains(neighborIndex)) {
                        continue;
                    }
                    queue.offer(neighborIndex);
                    visited.add(neighborIndex);
                }
            }
        }
        return 0;
    }
}

File: word-ladder/word-ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        adj_list = defaultdict(list)
        L = len(beginWord)
        
        for word in wordList:
            for i in range(L):
                pattern = word[:i] + "*" + word[i+1:]
                adj_list[pattern].append(word)
        
        print(adj_list)
        
        visited = set()
        q = Deque()
        q.append(beginWord)
        visited.add(beginWord)
        
        level = 1
        while q:
            size = len(q)
            for _ in range(size):
                word = q.popleft()
                for i in range(L):
                    pattern = word[:i] + "*" + word[i+1:]
                    for neighbor in adj_list[pattern]:
                        if neighbor == endWord:
                            return level + 1
                        if neighbor in visited:
                            continue

                        visited.add(neighbor)
                        q.append(neighbor)
            level += 1
        
        return 0